### ISSUE TITLE
Multiprocess doesn't work on Linux

### ISSUE THREAD

**Issue by lizhangecon** [tlamadon/pytwoway#45]
FEEstimator gets stuck in fitting when the sample size exceeds ~1 million observations. I am doing HE correction. After the "leverages batch" step, it stops at the "leverages" step. The progress bar is always 0%.

**Comment by lizhangecon** [comment]
It happens because the multiprocessing does not work. When I specify 'ncore': 1, it works well.

**Comment by tlamadon** [comment]
Do you mind telling us which platform you are using? Windows, mac, linux?

**Comment by adamoppenheimer** [comment]
It would also be helpful if you could please paste the code you are using.

In addition, you should see if you can run the code using simulated data, you can use the code from the benchmark:
```python
import numpy as np
import bipartitepandas as bpd

sim_params = bpd.sim_params({'n_workers': 500000, 'firm_size': 10, 'p_move': 0.05})
rng = np.random.default_rng(1234)

sim_data = bpd.SimBipartite(sim_params).simulate(rng)
```

I don't think the sample size is an issue, I've run the estimator with multiprocessing on simulated datasets larger than yours (for example in the benchmark) and it runs fine.

EDIT: please also try installing the package `multiprocess` by running `pip install multiprocess`, it is possible this could fix the issue you are having

**Comment by lizhangecon** [comment]
> Do you mind telling us which platform you are using? Windows, mac, linux?

I am using linux

**Comment by lizhangecon** [comment]
> It would also be helpful if you could please paste the code you are using.
> 
> In addition, you should see if you can run the code using simulated data, you can use the code from the benchmark:
> 
> ```python
> import numpy as np
> import bipartitepandas as bpd
> 
> sim_params = bpd.sim_params({'n_workers': 500000, 'firm_size': 10, 'p_move': 0.05})
> rng = np.random.default_rng(1234)
> 
> sim_data = bpd.SimBipartite(sim_params).simulate(rng)
> ```
> 
> I don't think the sample size is an issue, I've run the estimator with multiprocessing on simulated datasets larger than yours (for example in the benchmark) and it runs fine.
> 
> EDIT: please also try installing the package `multiprocess` by running `pip install multiprocess`, it is possible this could fix the issue you are having

Yes, I am using simulated data. Once the sample size is more than 1 million observations, multiprocessing is not working. I tried re-install the package multiprocess but it does not solve the problem.

**Comment by lizhangecon** [comment]
Below is the code using:
`import pandas as pd
import numpy as np
import pytwoway as tw
import bipartitepandas as bpd

sim_params = bpd.sim_params({'n_workers': 500000, 'firm_size': 10, 'p_move': 0.05})
rng = np.random.default_rng(1234)

sim_data = bpd.SimBipartite(sim_params).simulate(rng)

# FE
fe_params = tw.fe_params(
    {
        'he': True,
        'ncore': 8
    }
)
# Cleaning
clean_params = bpd.clean_params(
    {
        'connectedness': 'leave_out_spell',
        'collapse_at_connectedness_measure': True,
        'drop_single_stayers': True,
        'drop_returns': 'returners',
        'copy': False
    }
)

sim_data = bpd.SimBipartite(sim_params).simulate()

# Convert into BipartitePandas DataFrame
bdf = bpd.BipartiteDataFrame(sim_data)
# Clean and collapse
bdf = bdf.clean(clean_params)

# Initialize FE estimator
fe_estimator = tw.FEEstimator(bdf, fe_params)
# Fit FE estimator
fe_estimator.fit()`

**Comment by lizhangecon** [comment]
Below is the code I used:
```
import pandas as pd
import numpy as np
import pytwoway as tw
import bipartitepandas as bpd

sim_params = bpd.sim_params({'n_workers': 500000, 'firm_size': 10, 'p_move': 0.05})
rng = np.random.default_rng(1234)

sim_data = bpd.SimBipartite(sim_params).simulate(rng)

# FE
fe_params = tw.fe_params(
    {
        'he': True,
        'ncore': 8
    }
)
# Cleaning
clean_params = bpd.clean_params(
    {
        'connectedness': 'leave_out_spell',
        'collapse_at_connectedness_measure': True,
        'drop_single_stayers': True,
        'drop_returns': 'returners',
        'copy': False
    }
)

sim_data = bpd.SimBipartite(sim_params).simulate()

# Convert into BipartitePandas DataFrame
bdf = bpd.BipartiteDataFrame(sim_data)
# Clean and collapse
bdf = bdf.clean(clean_params)

# Initialize FE estimator
fe_estimator = tw.FEEstimator(bdf, fe_params)
# Fit FE estimator
fe_estimator.fit()
```

**Comment by adamoppenheimer** [comment]
The code you posted works for me (I am on Mac).

Could you please post the versions of bipartitepandas and pytwoway that you are using?

**Comment by lizhangecon** [comment]
> The code you posted works for me (I am on Mac).
> 
> Could you please post the versions of bipartitepandas and pytwoway that you are using?

bipartitepandas: 1.1.4
pytwoway: 0.3.21

**Comment by adamoppenheimer** [comment]
I think as @tlamadon mentioned this could be an os issue. Unfortunately I don't have Linux so can't test it myself, and I'm also taking classes so don't have time to do any real work on this for at least a month.

One last idea I have could be your python version: I'm running version 3.10.4, so if you're using a newer version, it's possible multiprocessing has changed and is not compatible with the code I wrote.

I hope in the meantime that using a single core is fast enough for what you need, otherwise you could always try using [LeaveOutTwoWay](https://github.com/rsaggio87/LeaveOutTwoWay/) (Matlab) or [VarianceComponentsHDFE](https://github.com/HighDimensionalEconLab/VarianceComponentsHDFE.jl) (Julia). I know the Matlab code has multiprocessing and I wouldn't be surprised if the Julia code also has multiprocessing. I would just recommend using bipartitepandas to compute the largest leave-one-out connected set, since as noted in the benchmark, the KSS code unnecessarily drops observations.

Feel free to post an update here if you manage to get multiprocessing to work, and please let us know what steps resolved the issue so that if anyone else encounters this in the future they can see what worked for you.

**Comment by adamoppenheimer** [comment]
I just quickly looked up issues with multiprocessing on Linux, maybe this could help?

https://stackoverflow.com/questions/55615981/why-does-my-python-multiprocessing-script-run-on-windows-but-not-on-linux

If you are using VSCode and switching to another IDE or Terminal doesn't work, I can try the `multiprocessing.set_start_method("spawn")` fix at some point after my classes are done.

**Comment by lizhangecon** [comment]
@adamoppenheimer Thanks a lot for your help! I tried LeaveOutTwoWay on Matlab, and it works. Also, for my data, single-core computation is acceptable, so just take your time to fix the problem.
